bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/RetryingReadRowsOperation.java

```diff
@@ -83,22 +83,31 @@ public void run() {
     // restart the clock.
     lastResponseMs = clock.currentTimeMillis();
     this.rowMerger = new RowMerger(rowObserver);
-    super.run();
+    synchronized (callLock) {
+      super.run();
+      // pre-fetch one more result, for performance reasons.
+      call.request(1);
+    }
   }
 
   /** {@inheritDoc} */
   @Override
   public void onMessage(ReadRowsResponse message) {
-    // Pre-fetch the next request
-    call.request(1);
     resetStatusBasedBackoff();
+    if (rowObserver instanceof ResponseQueueReader) {
```

Comment:
I think using grpc's pattern of disableAutoInboundFlowControl() & setOnReadyHandler() might make this a bit cleaner